### ISSUE TITLE
[proposal] add a `HitTestBehavior` that allow children and targets visually behind to ALWAYS receive events

### ISSUE THREAD

**Issue by iapicca** [flutter/flutter#75090]
opening the issue following @Hixie [comment](https://github.com/flutter/flutter/issues/47119#issuecomment-770311496)

## Use case

being aware of the ongoing effort to rectify the documentation on HitTestBehavior (https://github.com/flutter/flutter/issues/18450 , https://github.com/flutter/flutter/issues/74733)

the current definition of [HitTestBehavior](https://api.flutter.dev/flutter/rendering/HitTestBehavior-class.html) `translucent`
> Translucent targets both receive events within their bounds and permit targets visually behind them to also receive events.

some developers (included myself) have interpreted said definition as
**_ALWAYS_** `permit  targets visually behind them to also receive events.` 
which isn't the case as:

- https://github.com/flutter/flutter/issues/47119 does not apply to [GestureDetector](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html) in a [Stack](https://api.flutter.dev/flutter/widgets/Stack-class.html) 
- https://github.com/flutter/flutter/issues/74707#issuecomment-770281424 does not apply if the [Container](https://api.flutter.dev/flutter/widgets/Container-class.html) in a [Listener](https://api.flutter.dev/flutter/widgets/Listener-class.html) has a `child`
- https://github.com/flutter/flutter/issues/74733#issuecomment-767859584 does not apply to Widgets `inside` rather than `behind` which is admittedly unobvious (similar to the point above, but I'd argue it isn't the same)

since `the behaviors mentioned above are intended` and not `bugs`, 
please consider the following proposal

## Proposal

create a `HitTestBehavior` that ALWAYS ALLOW to receive events:
- children and `targets visually behind` alike
- does not discriminate between `Listener` and `GestureDetector`
- does not discriminate between `Widgets inside rather than behind`

**Comment by giorgio79** [comment]
I think I am facing this issue as well :) (with copy paste example at https://github.com/flutter/flutter/issues/75086 )

**Comment by iapicca** [comment]
> I think I am facing this issue as well :) (with copy paste example at #75086 )

@giorgio79 
the issue I'm reporting is`not a bug`, but a `feature request`

**Comment by nt4f04uNd** [comment]
@iapicca i think this still is formulated confusingly and doesn't offer the best solution
can you consider this? https://github.com/flutter/flutter/issues/47119#issuecomment-767623269
or should i open another issue in favor of this one

**Comment by iapicca** [comment]
> [#47119 (comment)](https://github.com/flutter/flutter/issues/47119#issuecomment-767623269)

@nt4f04uNd 
I'm not proposing any change to the current behavior of  `HitTestBehavior.translucent`
(if I'm told by the team <this behavior is intended> that's a good enough answer for me)

what I'm proposing is to add a new option to the existing `HitTestBehavior` (as for the title)

**Comment by iapicca** [comment]
> [#47119 (comment)](https://github.com/flutter/flutter/issues/47119#issuecomment-767623269)

@nt4f04uNd 
I'm not proposing any change to the current behavior of  `HitTestBehavior.translucent`
(if I'm told by the team <this behavior is intended> that's a good enough answer for me)

what I'm proposing is to add a new option to the existing `HitTestBehavior` (as for the title)

**Comment by nt4f04uNd** [comment]
@iapicca yes, i understand that. so that looks for me unnecessary, as there's no need to add new `HitTestBehavior`, we can just change how hit test is made with `Stack`, so that's why i'm referring to my comment

**Comment by iapicca** [comment]
> there's no need to add new `HitTestBehavior`, we can just change how hit test is made with `Stack`

@nt4f04uNd I do not agree with the first statement and I have doubts about the second

**Comment by Hixie** [comment]
> * #47119 does not apply to [GestureDetector](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html) in a [Stack](https://api.flutter.dev/flutter/widgets/Stack-class.html)

I don't understand this. That bug is actually about PageView (despite its current title), so it isn't illuminating. In what sense does it not apply to a GestureDetector in a Stack?

> * [#74707 (comment)](https://github.com/flutter/flutter/issues/74707#issuecomment-770281424) does not apply if the [Container](https://api.flutter.dev/flutter/widgets/Container-class.html) in a [Listener](https://api.flutter.dev/flutter/widgets/Listener-class.html) has a `child`

It definitely doesn't apply to Containers, since Containers don't have a HitTestBehavior (and indeed aren't a RenderObjectWidget, so can't do hit testing at all). I'm not sure what this is actually requesting. Is it suggesting a value of HitTestBehavior that should always add itself to the hit test result, but also always report itself as not hit? What is the use case?

> * [#74733 (comment)](https://github.com/flutter/flutter/issues/74733#issuecomment-767859584) does not apply to Widgets `inside` rather than `behind` which is admittedly unobvious
>   (similar to the point above, but I'd argue it isn't the same)

That link is just explaining what we do now. It isn't requesting anything or describing any problem (other than the documentation needing clarification).

> ## Proposal
> create a `HitTestBehavior` that ALWAYS ALLOW to receive events:
> 
> * children and `targets visually behind` alike
> * does not discriminate between `Listener` and `GestureDetector`
> * does not discriminate between `Widgets inside rather than behind`

That describes `HitTestBehavior.opaque`, as far as I can tell.

As per [my comment](https://github.com/flutter/flutter/issues/47119#issuecomment-770311496) in #47119, what we need is "a clear request, describing the problem that needs solving". This bug isn't that; it links to other bugs, it doesn't actually state a problem, it doesn't say _why_ the request is being made, it doesn't explain the use case.

**Comment by iapicca** [comment]
@Hixie 
I apologize for any confusion I might have created 

## 
> > * #47119 does not apply to [GestureDetector](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html) in a [Stack](https://api.flutter.dev/flutter/widgets/Stack-class.html)
>
> I don't understand this. That bug is actually about PageView (despite its current title), so it isn't illuminating. In what sense does it not apply to a GestureDetector in a Stack?

in my [comment](https://github.com/flutter/flutter/issues/47119#issuecomment-766883751) I provided 2 code samples
in which I believe is clear the difference of behavior between [GestureDetector](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html) and [Listener](https://api.flutter.dev/flutter/widgets/Listener-class.html)

<details>
<summary>`GestureDetector`code sample</summary>

```dart
  import 'package:flutter/material.dart';

  void main() => runApp(
        MaterialApp(
          home: Material(
            child: Stack(
              children: [for (var i = 0; i < 3; i++) MyGestureDetector(i)],
            ),
          ),
        ),
      );

  class MyGestureDetector extends StatelessWidget {
    MyGestureDetector(this.index) : super(key: ValueKey('MyListener: $index'));
    final int index;
    @override
    Widget build(BuildContext context) => Center(
          child: GestureDetector(
            behavior: HitTestBehavior.translucent,
            onTap: () => print('$index'),
            child: SizedBox.expand(),
          ),
        );
  }

```

</details>
 
<details>
<summary>`Listener` code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(
      MaterialApp(
        home: Material(
          child: Stack(
            children: [for (var i = 0; i < 3; i++) MyListener(i)],
          ),
        ),
      ),
    );

class MyListener extends StatelessWidget {
  MyListener(this.index) : super(key: ValueKey('MyListener: $index'));
  final int index;
  @override
  Widget build(BuildContext context) => Center(
        child: Listener(
          behavior: HitTestBehavior.translucent,
          onPointerDown: (_) => print('$index'),
          child: SizedBox.expand(),
        ),
      );
}
```

</details>

I understand that this behavior is intended, 
a `HitTestBehavior` that allow `GestureDetector` being "equally permissive" as a `Listener`
can be desirable for some developer

##

> 
> > * [#74707 (comment)](https://github.com/flutter/flutter/issues/74707#issuecomment-770281424) does not apply if the [Container](https://api.flutter.dev/flutter/widgets/Container-class.html) in a [Listener](https://api.flutter.dev/flutter/widgets/Listener-class.html) has a `child`
> 
> It definitely doesn't apply to Containers, since Containers don't have a HitTestBehavior (and indeed aren't a RenderObjectWidget, so can't do hit testing at all). I'm not sure what this is actually requesting. Is it suggesting a value of HitTestBehavior that should always add itself to the hit test result, but also always report itself as not hit? What is the use case?
> 

the difference in the behavior reported in the [issue](https://github.com/flutter/flutter/issues/74707) isn't about
a `Listener` having or not a `Container` as a child, as in the code sample provided the behavior change occurs
when the `Container` as a `child` against a `Container` without
 
I don't have the expertise to determine what is the best solution
(and clearly not the artistic skills either) but I try to represent what's on top of my head with this drawing
 
![image](https://user-images.githubusercontent.com/45637986/106617559-6bdb9b80-6577-11eb-9267-791216ba5182.png)

[PINK] --> first gesture detector 
- receive tap 
- trigger callback 
- `echo` dx, dy to all "listening" GestureDetector & Listener
[BLUE] --> container  (just stands there)
[YELLOW] -->  another gesture detector >
- get notified that at dx, dy whatever action has occured 
- triggers callbacks if any (regardless being obstructed)

a use case can be for example a `parallax effect` 
or scrolling / dragging without being obstructed by "obstacles"
